Add NDT test support to etl_worker

stephen-soltesz opened the pull request:
This change includes NDT support for the `etl_worker`.

To do this we are making deliberate trade-offs in performance by using an mutex to permit processing only one web100 snaplog file at a time.

As well, this change adds a "file size" histogram in order to collect the typical snaplog sizes in order to estimate whether our tmpfs filesystem size is adequate to process multiple snaplogs at a time (before modifying the library to run entirely in-memory).

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/m-lab/etl/40)
<!-- Reviewable:end -->

gfr10598 commented on metrics/metrics.go:
```diff
@@ -104,6 +105,29 @@ var (
+	FileSizeHistogram = prometheus.NewHistogram(
+		prometheus.HistogramOpts{
+			Name: "etl_web100_snaplog_file_size_bytes",
+			Help: "Size of individual snaplog files.",
```
Note: We should probably generalize this to work for any data type, but just snaplog is fine for now.

gfr10598 commented on parser/ndt.go:
```diff
@@ -2,52 +2,73 @@ package parser
 	// Write rawSnapLog to /mnt/tmpfs.
-	tmpFile := fmt.Sprintf("%s/%s", n.tmpDir, testName)
-	err := ioutil.WriteFile(tmpFile, rawSnapLog, 0644)
+	if !strings.HasSuffix(testName, "c2s_snaplog") && !strings.HasSuffix(testName, "s2c_snaplog") {
```
You need to accept s2c_snaplog too.

gfr10598 commented on web100/web100.go:
```diff
@@ -21,11 +21,16 @@ import (
 )
 
+var (
+	web100Lock sync.Mutex
```
Add a TODO to eliminate this lock.

gfr10598 commented on web100/web100.go:
```diff
@@ -100,28 +107,28 @@ func (w *Web100) logValues() (map[string]bigquery.Value, error) {
 	agent := C.web100_get_log_agent(log)
```
Would prefer a name other than log, just to avoid confusion with logging, both here and w.log above.

gfr10598 commented on cmd/etl_worker/etl_worker.go:
```diff
@@ -101,9 +101,10 @@ func getInserter(dt etl.DataType, fake bool) (etl.Inserter, error) {
-		// TODO - substitute appropriate parsers here and below.
-		return parser.NewTestParser(ins)
+		// TODO - eliminate need for "/mnt/tmpfs"
+		return parser.NewNDTParser(ins, etl.TableNames[etl.NDTData], "/mnt/tmpfs")
```
Can this tmp directory be specified within NewNDTParser?

stephen-soltesz commented:
Review status: 0 of 7 files reviewed at latest revision, 7 unresolved discussions, some commit checks broke.

---

*[cmd/etl_worker/etl_worker.go, line 105 at r1](https://reviewable.io:443/reviews/m-lab/etl/40#-Kip3Jgh-NozvHHdqBah-r1-105:-Kj-Yh__fSHnIeiwYRYV:b-896fix) ([raw file](https://github.com/m-lab/etl/blob/4dec8f956575452af9f2980d6836b51a87085303/cmd/etl_worker/etl_worker.go#L105)):*
<details><summary><i>Previously, gfr10598 (Gregory Russell) wrote…</i></summary><blockquote>

Can this tmp directory be specified within NewNDTParser?
</blockquote></details>

Done.

---

*[metrics/metrics.go, line 112 at r1](https://reviewable.io:443/reviews/m-lab/etl/40#-Kip3Jgh-NozvHHdqBai-r1-112:-Kj-ZfsJfhuDy2CY67bd:bol1346) ([raw file](https://github.com/m-lab/etl/blob/4dec8f956575452af9f2980d6836b51a87085303/metrics/metrics.go#L112)):*
<details><summary><i>Previously, gfr10598 (Gregory Russell) wrote…</i></summary><blockquote>

Note: We should probably generalize this to work for any data type, but just snaplog is fine for now.
</blockquote></details>

I agree. Added TODO.

---

*[metrics/metrics.go, line 115 at r1](https://reviewable.io:443/reviews/m-lab/etl/40#-Kip3Jgh-NozvHHdqBai-r1-115:-Kj-ZhsCmK6yQcoXFMOo:b-cssmea) ([raw file](https://github.com/m-lab/etl/blob/4dec8f956575452af9f2980d6836b51a87085303/metrics/metrics.go#L115)):*
<details><summary><i>Previously, gfr10598 (Gregory Russell) wrote…</i></summary><blockquote>

Suggest  500k, 800k, 900k, 1m, 1.2m 1.5m 2m,3,4,5,6,7,8,9, 10m, inf ?
</blockquote></details>

Updated to use the buckets from https://github.com/m-lab/etl/blob/gfr/metrics/metrics.go#L145

---

*[parser/ndt.go, line 33 at r1](https://reviewable.io:443/reviews/m-lab/etl/40#-Kip3Jgh-NozvHHdqBaj-r1-33:-Kj-Zs477OhG6dNbT0rY:b-v5543b) ([raw file](https://github.com/m-lab/etl/blob/4dec8f956575452af9f2980d6836b51a87085303/parser/ndt.go#L33)):*
<details><summary><i>Previously, gfr10598 (Gregory Russell) wrote…</i></summary><blockquote>

You need to accept s2c_snaplog too.
</blockquote></details>

This condition ignores all suffixes other than c2s_snaplog and s2c_snaplog. I'm imagining the second part of the condition didn't stand out?

---

*[parser/ndt_test.go, line 73 at r1](https://reviewable.io:443/reviews/m-lab/etl/40#-KipFSUR3NQfHkIU6L87-r1-73:-Kj-_q_ELetadf3EIvVx:b-5jzywn) ([raw file](https://github.com/m-lab/etl/blob/4dec8f956575452af9f2980d6836b51a87085303/parser/ndt_test.go#L73)):*
<details><summary><i>Previously, gfr10598 (Gregory Russell) wrote…</i></summary><blockquote>

Don't care for this name.  FakeInserter?
Alternatively, should we add functionality to FakeUploader instead?
</blockquote></details>

I peeked at the content of the `fake` package because that looked like the right place, but there is a lot of code there borrowed from the bigquery library and it wasn't apparent how to get the functionality I needed in the ndt test from the current fake interface.

I don't think we want two types named `FakeInserter`. Having a more general "fake uploader" package sounds like a good idea.

---

*[web100/web100.go, line 31 at r1](https://reviewable.io:443/reviews/m-lab/etl/40#-Kip3JgiChqifZLCr6wE-r1-31:-Kj-b9OLutImkqF8VF2X:b-896fix) ([raw file](https://github.com/m-lab/etl/blob/4dec8f956575452af9f2980d6836b51a87085303/web100/web100.go#L31)):*
<details><summary><i>Previously, gfr10598 (Gregory Russell) wrote…</i></summary><blockquote>

Add a TODO to eliminate this lock.
</blockquote></details>

Done.

---

*[web100/web100.go, line 107 at r1](https://reviewable.io:443/reviews/m-lab/etl/40#-Kip3JgiChqifZLCr6wE-r1-107:-Kj-bpgRvLr_H_rGGSWe:b-mlynfn) ([raw file](https://github.com/m-lab/etl/blob/4dec8f956575452af9f2980d6836b51a87085303/web100/web100.go#L107)):*
<details><summary><i>Previously, gfr10598 (Gregory Russell) wrote…</i></summary><blockquote>

Would prefer a name other than log, just to avoid confusion with logging, both here and w.log above.
</blockquote></details>

Okay. Changed to `snaplog` throughout.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/m-lab/etl/40)*
<!-- Sent from Reviewable.io -->

gfr10598 commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Review status: 0 of 7 files reviewed at latest revision, all discussions resolved, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/m-lab/etl/40#-:-Kj3GWZiPA7TE4c9U66U:bnfp4nl)*
<!-- Sent from Reviewable.io -->

stephen-soltesz commented:
The rebase with the current state of integration required a few final additions to update parse.go with the correct call the `NewNDTParser` and to simplify that, since it seems that the inserter is the authoritative source for the `TableName` now.
